### PR TITLE
feat: Added other parties and parents and respondents directions to CMO (FPLA-886)

### PR DESCRIPTION
**Pull request by Jamesnelson117**
**Before creating a pull request make sure that:**

- [x] commit messages are meaningful and follow good commit message guidelines
- [x] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-886

### Change description ###
CMO changes - 
- Adds parents and respondents directions along with dropdown key label
- Adds other parties directions along with dropdown key label

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

Screenshot 2019-11-27 at 11 13 28

Screenshot 2019-11-25 at 14 52 36

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java [hunk 1]
```diff
@@ -10,33 +10,49 @@
...
     }
 
     @PostMapping("/about-to-start")
     public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody CallbackRequest callbackrequest) {
         CaseDetails caseDetails = callbackrequest.getCaseDetails();
-        Map<String, Object> caseData = caseDetails.getData();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        if (!isNull(caseData.getCaseManagementOrder())) {
+            directionHelperService.sortDirectionsByAssignee(caseData.getCaseManagementOrder().getDirections())
```
Comment (amitk2281) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java]: Can we move it as private method to make it more readable?

Comment (amitk2281) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java]: Can we move it as private method to make it more readable?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/OrderType.java [hunk 2]
```diff
@@ -1,5 +1,6 @@
 package uk.gov.hmcts.reform.fpl.enums;
 
+
```
Comment (amitk2281) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/enums/OrderType.java]: Please remove extra line

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java [hunk 3]
```diff
@@ -52,4 +68,21 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody Cal
...
+        caseDetails.getData().put("otherPartiesDropdownKeyCMO",
+            draftCMOService.createOtherPartiesAssigneeDropdownKey(caseDetails));
+
+        caseDetails.getData().put("parentsAndRespondentsDropdownKeyCMO",
+            draftCMOService.createParentsAndRespondentAssigneeDropdownKey(caseDetails));
+
+        return caseDetails;
+    }
+
+    private CaseDetails removeExistingDirectionsFromCaseDetails(CaseDetails caseDetails) {
```
Comment (amitk2281) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java]: Just thinking do we need `FromCaseDetails` in method name as it seems bit obvious from Method signature. I think `removeExistingDirections` should be fine.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/Others.java [hunk 4]
```diff
@@ -1,16 +1,34 @@
...
 @AllArgsConstructor
 public class Others {
     private final Other firstOther;
     private final List<Element<Other>> additionalOthers;
+
+    @JsonIgnore
+    public List<Other> getAllOthers() {
+        ImmutableList.Builder<Other> builder = ImmutableList.builder();
+
+        if (firstOther != null) {
```
Comment (amitk2281) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/Others.java]: I think we can move it in single line `if (firstOther != null) { builder.add(firstOther); }`

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java [hunk 5]
```diff
@@ -6,31 +6,41 @@
...
 
 @Service
 public class DraftCMOService {
     private final ObjectMapper mapper;
     private final DateFormatterService dateFormatterService;
+    private final DirectionHelperService directionHelperService;
 
     @Autowired
-    public DraftCMOService(DateFormatterService dateFormatterService, ObjectMapper mapper) {
+    public DraftCMOService(DateFormatterService dateFormatterService, ObjectMapper mapper,
```
Comment (amitk2281) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java]: just bit formatting, can move ObjectMapper to next line

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/Others.java (hunk 4 above)
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/Others.java]: Linter does not like this :(

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java [hunk 6]
```diff
@@ -70,6 +80,53 @@ private void prePopulateHearingDateSelection(List<Element<HearingBooking>> heari
...
+        }
+
+        return stringBuilder.toString().stripTrailing();
+    }
+
+    private String getRespondentsFullName(RespondentParty respondentParty) {
+        String firstName = defaultIfNull(respondentParty.getFirstName(), "");
+        String lastName = defaultIfNull(respondentParty.getLastName(), "");
+
+        return firstName + " " + lastName;
```
Comment (amitk2281) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java]: Can we use StringBuilder to concatenate string here?

Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java]: I can use string.format. I think stringBuilder is a bit of overkill for this.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseManagementOrder.java [hunk 7]
```diff
@@ -14,6 +16,7 @@
 public class CaseManagementOrder {
     private final String hearingDate;
     private final UUID hearingDateId;
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseManagementOrder.java]: This needs removed!

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOControllerTest.java [hunk 8]
```diff
@@ -56,8 +58,11 @@
     private ObjectMapper mapper;
 
     @Test
-    void aboutToStartCallbackShouldPopulateHearingDatesList() throws Exception {
-        Map<String, Object> data = ImmutableMap.of("hearingDetails", hearingDetails);
+    void aboutToStartCallbackShouldPopulateHearingDatesListAndSetAssigneeDropdownKeys() throws Exception {
```
Comment (joez-f) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOControllerTest.java]: Does this test do too much? Would it make more sense to split into two tests? Keep what we had before (populate hearing dates list) and then another test for setting assignee dropdown keys, they seem to be unrelated

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseManagementOrder.java (hunk 7 above)
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseManagementOrder.java]: Agree

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOControllerTest.java (hunk 8 above)
Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOControllerTest.java]: No because we would be doing the same setup in both tests just separating out the assertions. It's better to have one the int test with multiple assertions.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java [hunk 9]
```diff
@@ -70,6 +81,53 @@ private void prePopulateHearingDateSelection(List<Element<HearingBooking>> heari
...
+        StringBuilder stringBuilder = new StringBuilder();
+
+        if (isNotEmpty(caseData.getRespondents1())) {
+            for (int i = 0; i < caseData.getRespondents1().size(); i++) {
+                RespondentParty respondentParty = caseData.getRespondents1().get(i).getValue().getParty();
+
+                String key = String.format("Respondent %d - %s", i + 1, getRespondentsFullName(respondentParty));
+                stringBuilder.append(key).append("\n\n");
+            }
+        }
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java]: how about:
```
        List<Element<Respondent>> respondentsList = caseData.getRespondents1();
        if (isNotEmpty(respondentsList)) {
            for (Element<Respondent> respondent : respondentsList) {
                String key = String.format("Respondent %d - %s", respondentsList.size() + 1,
                    getRespondentsFullName(respondent.getValue().getParty()));
                stringBuilder.append(key).append("\n\n");
            }
```
bit more readable perhaps?

File: ccd-definition/CaseEventToComplexTypes/draft-CMO/otherPartiesDirectionsCustom.json [hunk 10]
```diff
@@ -0,0 +1,39 @@
+[
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "OtherPartiesDirectionsCustom",
+    "CaseEventID": "draftCMO",
+    "CaseFieldID": "otherPartiesDirectionsCustom",
+    "ListElementCode": "directionType",
+    "EventElementLabel": " ",
```
Comment (Jamesnelson117) [inline ccd-definition/CaseEventToComplexTypes/draft-CMO/otherPartiesDirectionsCustom.json]: Remove

File: ccd-definition/CaseEventToComplexTypes/draft-CMO/parentsAndRespondentsCustom.json [hunk 11]
```diff
@@ -0,0 +1,39 @@
+[
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "ParentsAndRespondentsCustom",
+    "CaseEventID": "draftCMO",
+    "CaseFieldID": "respondentDirectionsCustom",
+    "ListElementCode": "directionType",
+    "EventElementLabel": " ",
```
Comment (Jamesnelson117) [inline ccd-definition/CaseEventToComplexTypes/draft-CMO/parentsAndRespondentsCustom.json]: Remove

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java [hunk 12]
```diff
@@ -70,6 +81,53 @@ private void prePopulateHearingDateSelection(List<Element<HearingBooking>> heari
...
+                }
+
+                stringBuilder.append(key).append("\n\n");
+            }
+        }
+
+        return stringBuilder.toString().stripTrailing();
+    }
+
+    private String getRespondentsFullName(RespondentParty respondentParty) {
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java]: Method name should be singular

File: e2e/fragments/draftDirections.js [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+          year: `#${party}_${index}_dateToBeCompletedBy-year`,
+          hour: `#${party}_${index}_dateToBeCompletedBy-hour`,
+          minute: `#${party}_${index}_dateToBeCompletedBy-minute`,
+          second: `#${party}_${index}_dateToBeCompletedBy-second`,
+        },
+      },
+    };
+  },
+
+  enterDate(party, direction, index = 0) {
```
Comment (geek-soft) [inline e2e/fragments/draftDirections.js]: Passing while direction to a matched that requires only one property of a direction is bad idea for many reasons. Always pass the minimum needed. It will make methods more meaningful / easier to understand. Passing more than needed suggests that much more is needed / method does much more than it is in reality.

File: e2e/fragments/draftDirections.js [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    I.fillField(this.fields(party, index).direction.dueDate.day, direction.dueDate.day);
+    I.fillField(this.fields(party, index).direction.dueDate.month, direction.dueDate.month);
+    I.fillField(this.fields(party, index).direction.dueDate.year, direction.dueDate.year);
+    I.fillField(this.fields(party, index).direction.dueDate.hour, direction.dueDate.hour);
+    I.fillField(this.fields(party, index).direction.dueDate.minute, direction.dueDate.minute);
+    I.fillField(this.fields(party, index).direction.dueDate.second, direction.dueDate.second);
+  },
+
+  enterTitleAndDescription(party, direction, index = 0) {
```
Comment (geek-soft) [inline e2e/fragments/draftDirections.js]: Same here, you need to pass title and description not whole direction.

File: e2e/pages/events/draftCaseManagementOrderEvent.page.js [hunk 15]
```diff
@@ -1,4 +1,5 @@
 const { I } = inject();
+const draftDirections = require('../../fragments/draftDirections');
```
Comment (geek-soft) [inline e2e/pages/events/draftCaseManagementOrderEvent.page.js]: What is `draftDirections`? Is that an object containing directions draft? I would not have doubts if constant was named right way.

File: e2e/pages/events/draftCaseManagementOrderEvent.page.js [hunk 16]
```diff
@@ -46,6 +53,18 @@ module.exports = {
     I.fillField(this.fields.schedule.partiesPositions, schedule.partiesPositions);
   },
 
+  async enterDirection(direction) {
+    await I.addAnotherElementToCollection();
+    await draftDirections.enterTitleAndDescription('respondentDirectionsCustom', direction);
```
Comment (geek-soft) [inline e2e/pages/events/draftCaseManagementOrderEvent.page.js]: Methods in `draftDirections` are not async so no need to await for a result. Remove unnecessary keyword. It is a shame that JS is so forgiving.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java (hunk 1 above)
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java]: Not a problem with that PR directly but `sortDirectionsByAssignee` method does not sort, it groups by assignee.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java [hunk 17]
```diff
@@ -50,6 +61,67 @@ public DynamicList getHearingDateDynamicList(CaseDetails caseDetails) {
...
+
+        StringBuilder stringBuilder = new StringBuilder();
+
+        if (isNotEmpty(caseData.getOthers())) {
+            for (int i = 0; i < caseData.getOthers().getAllOthers().size(); i++) {
+                Other other = caseData.getOthers().getAllOthers().get(i);
+                String key;
+
+                if (i == 0) {
+                    key = String.format("Person %d - %s", i + 1, defaultIfNull(other.getName(), ""));
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java]: Can person name be null or empty? If so it will render `Person 1 - ` which does not look great. Dash should be added only when name is available.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/DraftCMOServiceTest.java [hunk 18]
```diff
@@ -92,6 +95,50 @@ void shouldReturnCaseManagementOrderWhenProvidedCaseDetails() {
             formatLocalDateToMediumStyle(5));
     }
 
+    @Test
+    void shouldFormatRespondentsIntoKeyWhenRespondentsArePresent() {
+        CaseDetails caseDetails = CaseDetails.builder()
```
Comment (geek-soft) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/DraftCMOServiceTest.java]: Nasty consequence of accepting `CaseDetails` everywhere is that it needs to be created in test, quite pointless. If correctly implemented method would just take list of respondents. It would not care about case details nor case data. It would not have to know how to retrieve relevant details from 'context' objects.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/DraftCMOServiceTest.java [hunk 19]
```diff
@@ -92,6 +95,50 @@ void shouldReturnCaseManagementOrderWhenProvidedCaseDetails() {
...
+    void shouldReturnEmptyStringWhenRespondentsAreNotPresent() {
+        CaseDetails caseDetails = CaseDetails.builder().data(ImmutableMap.of()).build();
+
+        String respondentsKey = draftCMOService.createRespondentAssigneeDropdownKey(caseDetails);
+
+        assertThat(respondentsKey).isEqualTo("");
+    }
+
+    @Test
+    void shouldFormatOthersIntoKeyWhenOthersArePresent() {
```
Comment (geek-soft) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/DraftCMOServiceTest.java]: What about test case where name is null or empty?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java (hunk 17 above)
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java]: It can, and there is also no safety guards on respondents fullName being empty string. I'd thought about this and asked Moray about placeholders if fullName was empty in both cases. To which I was pointed to a ticket https://tools.hmcts.net/jira/browse/FPLA-962 created just for the validation on this. 

It would be difficult to add validation without partially completing the above ticket.

File: ccd-definition/CaseField.json [hunk 20]
```diff
@@ -670,5 +670,37 @@
...
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "ID": "respondentsDirectionLabelCMO",
+    "Label": "## For the parents or respondents",
+    "FieldType": "Label",
+    "SecurityClassification": "Public"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "ID": "respondentsDropdownKeyCMO",
```
Comment (geek-soft) [inline ccd-definition/CaseField.json]: Why the name contains 'key' word? What does it mean?

File: ccd-definition/ComplexTypes/2_direction.json [hunk 21]
```diff
@@ -62,6 +62,28 @@
...
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "Direction",
+    "ListElementCode": "parentsAndRespondentsAssignee",
+    "FieldType": "FixedList",
+    "FieldTypeParameter": "ParentsAndRespondentsDirectionAssignee",
+    "ElementLabel": "Assignee",
+    "SecurityClassification": "Public",
+    "FieldShowCondition": "directionNeeded!=\"No\"",
+    "Comment": "This field is used for capturing parents and respondents assignee on CMO"
```
Comment (geek-soft) [inline ccd-definition/ComplexTypes/2_direction.json]: Comment? Is that kept in XLS file or just transient value in JSON for own own benefits?

Comment (lewisbirks) [reply ccd-definition/ComplexTypes/2_direction.json]: It is a transient value, just added to explain the field or whatever it is attached to

File: ccd-definition/CaseField.json (hunk 20 above)
Comment (Jamesnelson117) [reply ccd-definition/CaseField.json]: Screenshot 2019-11-25 at 14 49 16

It's the key of the assignee dropdown in directions.

File: e2e/pages/events/draftCaseManagementOrderEvent.page.js [hunk 22]
```diff
@@ -20,6 +21,12 @@ module.exports = {
       keyIssues: '#schedule_keyIssues',
       partiesPositions: '#schedule_partiesPositions',
     },
+    respondentDirectionsCustom: {
```
Comment (geek-soft) [inline e2e/pages/events/draftCaseManagementOrderEvent.page.js]: Why is that defined here? Shouldn't that be part of `draftDirections` fragment?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOControllerTest.java (hunk 8 above)
Comment (geek-soft) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOControllerTest.java]: Please move it to operate test case. Test should do one thing and do it well. If method name contains words like 'and' it simple indication that test is checking too many things.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java [hunk 23]
```diff
@@ -10,33 +10,49 @@
...
+        this.directionHelperService = directionHelperService;
     }
 
     @PostMapping("/about-to-start")
     public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody CallbackRequest callbackrequest) {
         CaseDetails caseDetails = callbackrequest.getCaseDetails();
-        Map<String, Object> caseData = caseDetails.getData();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        if (!isNull(caseData.getCaseManagementOrder())) {
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java]: Logic behind that if/else is never tested. How do you know it works?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java [hunk 24]
```diff
@@ -10,33 +10,49 @@
...
+            directionHelperService.sortDirectionsByAssignee(caseData.getCaseManagementOrder().getDirections())
+                .forEach((key, value) -> caseDetails.getData().put(key.getValue(), value));
+        } else {
+            removeExistingCustomDirections(caseDetails);
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java]: Why you need to remove custom directions when CMO is not there? Why would custom directions be in case data if CMO is not there?

File: e2e/pages/events/draftCaseManagementOrderEvent.page.js (hunk 22 above)
Comment (Jamesnelson117) [reply e2e/pages/events/draftCaseManagementOrderEvent.page.js]: No because only the respondentsDirections and otherDirections in CMO will have this dropdown.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java [hunk 25]
```diff
@@ -10,33 +10,49 @@
...
         this.mapper = mapper;
         this.draftCMOService = draftCMOService;
+        this.directionHelperService = directionHelperService;
     }
 
     @PostMapping("/about-to-start")
     public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody CallbackRequest callbackrequest) {
         CaseDetails caseDetails = callbackrequest.getCaseDetails();
-        Map<String, Object> caseData = caseDetails.getData();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java]: That is so wired. There is a JSON>Java conversion to allow easier data read but all writes happen on maps. 👀

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java (hunk 24 above)
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java]: Because customDirection are shared between SDO and CMO

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/Direction.java [hunk 26]
```diff
@@ -20,6 +22,8 @@
     private String directionText;
     private final String status;
     private DirectionAssignee assignee;
+    private ParentsAndRespondentsDirectionAssignee parentsAndRespondentsAssignee;
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/Direction.java]: In healthy environment it makes more sense to have it in more specialised class. Here `Direction` class became one big bag of properties not helping to understand which one are needed and when :(

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java (hunk 23 above)
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java]: So it's tested in FPLA-657, did not want to repeat something I was going to merge in before story completion.

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOControllerTest.java (hunk 8 above)
Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOControllerTest.java]: I agree but it would seem there is conflicting opinions in the team!

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/Others.java [hunk 27]
```diff
@@ -1,16 +1,34 @@
...
+import java.util.stream.Collectors;
 
 @Data
 @Builder
 @AllArgsConstructor
 public class Others {
     private final Other firstOther;
     private final List<Element<Other>> additionalOthers;
+
+    @JsonIgnore
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/Others.java]: :( Keeping these first other / additional others is depressing. Better would be to use either JsonCreator if only one deserialisation is needed or pair serialisers / deserialisers.

I guess not for now.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java [hunk 28]
```diff
@@ -318,7 +318,14 @@ private String booleanToYesOrNo(boolean value) {
...
 
-    private List<Element<Direction>> assignCustomDirections(List<Element<Direction>> directions,
+    /**
+     * Iterates over a list of directions and sets properties assignee, custom and readOnly.
+     *
+     * @param directions  a list of directions.
+     * @param assignee    the assignee of the directions to be returned.
+     * @return A list of custom directions.
+     */
+    public List<Element<Direction>> assignCustomDirections(List<Element<Direction>> directions,
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DirectionHelperService.java]: Jesus...

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java (hunk 17 above)
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java]: Caught up with Moray, if firstName does not exist it will read 'BLANK - Please complete'

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/DraftCMOServiceTest.java (hunk 19 above)
Comment (Jamesnelson117) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/DraftCMOServiceTest.java]: Added additional test to assert if an other contains no full name then the empty placeholder of "BLANK - please complete" will appear.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/Others.java (hunk 27 above)
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/Others.java]: We had plans to migrate this but unfortunately never happened.

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOControllerTest.java (hunk 8 above)
Comment (geek-soft) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOControllerTest.java]: Conflicting and bad.

Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOControllerTest.java]: What if we re-considered the test name `aboutToStartCallbackShouldPrepareCaseForCMO`

Comment (lewisbirks) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOControllerTest.java]: That would be useful seeing as my story is going to be doing more of the same when I get round to editing that test

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java (hunk 23 above)
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java]: Move this out to a service method to test it properly.
